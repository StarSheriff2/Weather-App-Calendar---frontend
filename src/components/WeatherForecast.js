import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import { useDispatch } from 'react-redux';
import openWeatherApiService from '../services/openWeatherApi.service';
import { setMessage } from '../slices/message';

const WeatherForecast = ({
  coordinates, dateTime, dateDiff, currentReminder,
}) => {
  const dispatch = useDispatch();
  const [weatherData, setWeatherData] = useState();
  const [error, setError] = useState();
  const [lat, lon] = coordinates.split(', ');

  const fetchWeatherData = async () => {
    if (currentReminder) {
      const { data } = await openWeatherApiService.getCurrentWeather({ lat, lon });
      const { weather, main } = data;
      return [weather[0].icon, Math.round(main.temp)];
    }
    const { data } = await openWeatherApiService.getDailyWeather({ lat, lon });
    const { daily: tempByDate } = data;
    const { weather } = tempByDate[dateDiff];
    const { icon } = weather[0];
    const reminderHour = dateTime.getHours();
    const { temp } = tempByDate[dateDiff];

    if (reminderHour >= 5 && reminderHour < 12) {
      return [icon, Math.round(temp.morn)];
    } if (reminderHour >= 12 && reminderHour < 17) {
      return [icon, Math.round(temp.day)];
    } if (reminderHour >= 17 && reminderHour < 19) {
      return [icon, Math.round(temp.eve)];
    }
    return [icon, Math.round(temp.night)];
  };

  useEffect(async () => {
    try {
      const [icon, temp] = await fetchWeatherData();
      setWeatherData([icon, temp]);
    } catch (error) {
      dispatch(setMessage('Problems fetching weather data. Try reloading the page.'));
      setError('Network error');
    }

    return () => {
      setWeatherData();
    };
  }, []);

  if (!weatherData && !error) {
    return (<span className="spinner-border spinner-border-sm" />);
  }

  if (error) {
    return (<i className="fas fa-exclamation-circle text-warning" />);
  }

  return (
    weatherData && (
      <div className="d-flex flex-column justify-content-between align-items-center">
        <span>{`${weatherData[1]}Â°`}</span>
        <span><img src={`http://openweathermap.org/img/wn/${weatherData[0]}@2x.png`} alt="weather icon" width="60" /></span>
      </div>
    )
  );
};

WeatherForecast.propTypes = {
  coordinates: PropTypes.string.isRequired,
  dateTime: PropTypes.shape(
    null,
  ).isRequired,
  dateDiff: PropTypes.number.isRequired,
  currentReminder: PropTypes.bool,
};

WeatherForecast.defaultProps = {
  currentReminder: false,
};

export default WeatherForecast;
