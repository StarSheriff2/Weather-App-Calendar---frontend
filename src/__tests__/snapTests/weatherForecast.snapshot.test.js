import React from 'react';
import { cleanup } from '@testing-library/react';
import { BrowserRouter as Router } from 'react-router-dom';
import renderWithRedux from '../../tests/test-utils';
import WeatherForecast from '../../components/WeatherForecast';

afterEach(cleanup);

describe('WeatherForecast', () => {
  test('Renders the Component', async () => {
    const coordinates = '31.438327740769708, -69.93035045104288';
    const reminderTime = new Date('2022-02-03T05:39:00-06:00');
    const dateDiff = 0;

    const weatherForecast = renderWithRedux(
      <Router>
        <WeatherForecast
          coordinates={coordinates}
          dateTime={reminderTime}
          dateDiff={dateDiff}
        />
      </Router>,
    ).toJSON();

    expect(weatherForecast).toMatchSnapshot();
  });
});
