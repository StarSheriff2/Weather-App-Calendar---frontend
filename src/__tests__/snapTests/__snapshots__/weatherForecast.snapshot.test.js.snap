// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WeatherForecast Renders the Component 1`] = `
<span
  className="spinner-border spinner-border-sm"
/>
`;
